`flutter build apk` fails

### Describe the bug
Tried to update flutter to latest version and cannot seem to build apk. Looks quite similar to #95071, but flutter run does not work either. 

```

💪 Building with sound null safety 💪

Running Gradle task 'assembleRelease'...
Target android_aot_release_android-arm failed: ProcessException: Failed to find "/nix/store/d8bw0xsa3jja3z0hzbl8im78rl2x23z5-flutter-3.7.12-unwrapped/bin/cache/artifacts/engine/android-arm-release/linux-x64/gen_snapshot" in the search path.
  Command: /nix/store/d8bw0xsa3jja3z0hzbl8im78rl2x23z5-flutter-3.7.12-unwrapped/bin/cache/artifacts/engine/android-arm-release/linux-x64/gen_snapshot
Target android_aot_release_android-arm64 failed: ProcessException: Failed to find "/nix/store/d8bw0xsa3jja3z0hzbl8im78rl2x23z5-flutter-3.7.12-unwrapped/bin/cache/artifacts/engine/android-arm64-release/linux-x64/gen_snapshot" in the search path.
  Command: /nix/store/d8bw0xsa3jja3z0hzbl8im78rl2x23z5-flutter-3.7.12-unwrapped/bin/cache/artifacts/engine/android-arm64-release/linux-x64/gen_snapshot
Target android_aot_release_android-x64 failed: ProcessException: Failed to find "/nix/store/d8bw0xsa3jja3z0hzbl8im78rl2x23z5-flutter-3.7.12-unwrapped/bin/cache/artifacts/engine/android-x64-release/linux-x64/gen_snapshot" in the search path.
  Command: /nix/store/d8bw0xsa3jja3z0hzbl8im78rl2x23z5-flutter-3.7.12-unwrapped/bin/cache/artifacts/engine/android-x64-release/linux-x64/gen_snapshot


FAILURE: Build failed with an exception.

* Where:
Script '/nix/store/d8bw0xsa3jja3z0hzbl8im78rl2x23z5-flutter-3.7.12-unwrapped/packages/flutter_tools/gradle/flutter.gradle' line: 1157

* What went wrong:
Execution failed for task ':app:compileFlutterBuildRelease'.
> Process 'command '/nix/store/d8bw0xsa3jja3z0hzbl8im78rl2x23z5-flutter-3.7.12-unwrapped/bin/flutter'' finished with non-zero exit value 1

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 2s
Running Gradle task 'assembleRelease'...                            3.1s
Gradle task assembleRelease failed with exit code 1
```


### Steps To Reproduce
Steps to reproduce the behavior:
1. Clone: https://github.com/babariviere/flutter-nix-hello-world
2. `nix develop` and `flutter build apk`. Everything builds
3. Exit shell and `nix flake update` 
4. Comment out `emulatorVersion` and add "33" to `platformVersion` in `./nix/android.nix`. 
5. `nix develop` and `flutter clean && flutter build apk`. Error returned. 

### Additional context
Error says that it's looking for `/nix/store/d8bw0xsa3jja3z0hzbl8im78rl2x23z5-flutter-3.7.12-unwrapped/bin/cache/artifacts/engine/android-arm-release/linux-x64/gen_snapshot`, this does not seem to exist. However `gen_snapshot` seems to be placed one dir up: `/nix/store/d8bw0xsa3jja3z0hzbl8im78rl2x23z5-flutter-3.7.12-unwrapped/bin/cache/artifacts/engine/android-arm-release/gen_snapshot`.

### Notify maintainers
@babariviere @hacker1024 @gilice @ericdallo 

### Metadata
```console
 - system: `"x86_64-linux"`
 - host os: `Linux 6.1.26, NixOS, 23.05 (Stoat), 23.05.20230430.da45bf6`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.15.0`
 - nixpkgs: `/nix/store/3rfn91lqch3wchpgqamxl8sazbpnxbm9-source`
```

Thanks for the quick response! 

I tried to test the pull request by adding
`inputs.nixpkgs.url = "github:nixos/nixpkgs?rev=8fca20ca1c1849a37b1d1af8a0b14fd763a192c4";`
to my flake.nix. This should be a way to test it right? Another issue then arises: 

```
💪 Building with sound null safety 💪

Running Gradle task 'assembleRelease'...

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':app:mergeReleaseResources'.
> Multiple task action failures occurred:
   > A failure occurred while executing com.android.build.gradle.internal.res.Aapt2CompileRunnable
      > AAPT2 aapt2-7.1.2-7984345-linux Daemon #0: Daemon startup failed
        This should not happen under normal circumstances, please file an issue if it does.
   > A failure occurred while executing com.android.build.gradle.internal.res.Aapt2CompileRunnable
      > AAPT2 aapt2-7.1.2-7984345-linux Daemon #1: Daemon startup failed
        This should not happen under normal circumstances, please file an issue if it does.
   > A failure occurred while executing com.android.build.gradle.internal.res.Aapt2CompileRunnable
      > AAPT2 aapt2-7.1.2-7984345-linux Daemon #4: Daemon startup failed
        This should not happen under normal circumstances, please file an issue if it does.
   > A failure occurred while executing com.android.build.gradle.internal.res.Aapt2CompileRunnable
      > AAPT2 aapt2-7.1.2-7984345-linux Daemon #3: Daemon startup failed
        This should not happen under normal circumstances, please file an issue if it does.
   > A failure occurred while executing com.android.build.gradle.internal.res.Aapt2CompileRunnable
      > AAPT2 aapt2-7.1.2-7984345-linux Daemon #2: Daemon startup failed
        This should not happen under normal circumstances, please file an issue if it does.
   > A failure occurred while executing com.android.build.gradle.internal.res.Aapt2CompileRunnable
      > AAPT2 aapt2-7.1.2-7984345-linux Daemon #6: Daemon startup failed
        This should not happen under normal circumstances, please file an issue if it does.
   > A failure occurred while executing com.android.build.gradle.internal.res.Aapt2CompileRunnable
      > AAPT2 aapt2-7.1.2-7984345-linux Daemon #5: Daemon startup failed
        This should not happen under normal circumstances, please file an issue if it does.

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 10s
Running Gradle task 'assembleRelease'...                           10.6s
Gradle task assembleRelease failed with exit code 1
```

--- a/pkgs/development/compilers/flutter/engine-artifacts/default.nix
+++ b/pkgs/development/compilers/flutter/engine-artifacts/default.nix
@@ -80,7 +80,7 @@ let
       };
     };
 
-  mkArtifactDerivation = { platform ? null, variant ? null, archive, ... }@args:
+  mkArtifactDerivation = { platform ? null, variant ? null, subdirectory ? null, archive, ... }@args:
     let
       artifactDirectory = if platform == null then null else "${platform}${lib.optionalString (variant != null) "-${variant}"}";
       archiveBasename = lib.removeSuffix ".${(lib.last (lib.splitString "." archive))}" archive;
@@ -97,7 +97,14 @@ let
 
       nativeBuildInputs = [ autoPatchelfHook ];
 
-      installPhase = "cp -r . $out";
+      installPhase =
+        let
+          destination = "$out/${if subdirectory == true then archiveBasename else if subdirectory != null then subdirectory else "."}";
+        in
+        ''
+          mkdir -p "${destination}"
+          cp -r . "${destination}"
+        '';
     } // args);
 
   artifactDerivations = {

--- a/pkgs/development/compilers/flutter/engine-artifacts/default.nix
+++ b/pkgs/development/compilers/flutter/engine-artifacts/default.nix
@@ -30,7 +30,7 @@ let
                 variants = lib.genAttrs [ "profile" "release" ]
                   (variant: [
                     { archive = "artifacts.zip"; }
-                    { archive = "${lib.toLower hostPlatform.uname.system}-x64.zip"; }
+                    { subdirectory = true; archive = "${lib.toLower hostPlatform.uname.system}-x64.zip"; }
                   ]);
               })) //
           {

